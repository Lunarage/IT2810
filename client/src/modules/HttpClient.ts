import fetch, { Response } from "node-fetch";
import { Like, Movie, User } from "../types/DatabaseTypes";

/**
 * A custom error for response codes that are not 2xx
 *
 * @extends Error
 */
class HttpRequestError extends Error {
    statusCode: number;
    constructor(statusCode: number, message: string) {
        super(message);
        this.name = "HttpRequestError";
        this.message = message;
        this.statusCode = statusCode;
    }
}

// Object containing search parameters
// See README for more information
type SearchParameters = {
    title?: string;
    titleType?: string;
    minYear?: number;
    maxYear?: number;
    username?: string;
    page?: number;
    orderBy?: string;
    orderDir?: string;
};

class HttpClient {
    private baseURL: string;

    constructor(baseURL: string) {
        this.baseURL = baseURL;
    }

    // Helper Functions
    /**
     * Checks the response code and throws an error if it is not 2xx.
     * Otherwise it just returns the response.
     *
     * @param {Response} response - Object that represents the HTTP response
     * @throws {HttpRequestError} - Error Object
     */
    public checkStatus(response: Response) {
        if (response.ok) {
            return response;
        } else {
            throw new HttpRequestError(response.status, response.statusText);
        }
    }

    // Generic Functions
    /**
     * Sends a get request to the specified url
     * and returns a promise of a result
     * with the expected type.
     * Throws an error if the url can't be reached
     * or the response code is not 2xx
     *
     * @template T - The expected type of the response
     * @param {string} url - The url to send the request to
     * @return {Promise<T>} The promise of a response
     */
    public get<T>(url: string): Promise<T> {
        return fetch(url, { method: "GET" })
            .then(this.checkStatus)
            .then((response) => response.json())
            .then((response) => {
                return response as T;
            });
    }

    /**
     * Sends a delete request to the specified url
     * and returns a promise of a result
     * with the expected type.
     * Throws an error if the url can't be reached
     * or the response code is not 2xx
     *
     * @template T - The expected type of the response
     * @param {string} url - The url to send the request to
     * @return {Promise<T>} The promise of a response
     */
    public delete<T>(url: string): Promise<T> {
        return fetch(url, { method: "DELETE" })
            .then(this.checkStatus)
            .then((response) => response.json())
            .then((response) => {
                return response as T;
            });
    }

    /**
     * Sends a put request to the specified url
     * and returns a promise of a result
     * with the expected type.
     * Throws an error if the url can't be reached
     * or the response code is not 2xx
     *
     * @template T - The expected type of the response
     * @param {string} url - The url to send the request to
     * @return {Promise<T>} The promise of a response
     */
    public put<T>(url: string): Promise<T> {
        return fetch(url, { method: "PUT" })
            .then(this.checkStatus)
            .then((response) => response.json())
            .then((response) => {
                return response as T;
            });
    }

    // Specific functions

    /**
     * Gets a spesific movie.
     *
     * @param {string} movieId - The id of the movie. e.g "tt9655334"
     * @param {string} [username] - Optionally supply a username to get liked status
     * @return {Promise<Movie>} The promise of a Movie
     */
    public getMovie(movieId: string, username?: string): Promise<Movie> {
        let searchURL = this.baseURL + "/movie/" + movieId;
        if (username) {
            searchURL += "?username=" + username;
        }
        return this.get<Movie[]>(searchURL).then((response) => {
            return (response as Movie[])[0];
        });
    }

    /**
     * Searches for Movies
     *
     * @param {SearchParameters} args - Arguments for the search. See README for more information.
     * @return {Promise<Movie[]>} The promise of a list of movies
     */
    public searchMovies(args: SearchParameters): Promise<Movie[]> {
        // Build query from arguments
        let searchURL = this.baseURL + "/movie";
        let delimiter = "?";
        if (args.title) {
            searchURL += delimiter + "title=" + args.title;
            delimiter = "&";
        }
        if (args.titleType) {
            searchURL += delimiter + "titleType=" + args.titleType;
            delimiter = "&";
        }
        if (args.minYear) {
            searchURL += delimiter + "minYear=" + args.minYear;
            delimiter = "&";
        }
        if (args.maxYear) {
            searchURL += delimiter + "maxYear=" + args.maxYear;
            delimiter = "&";
        }
        if (args.orderBy) {
            searchURL += delimiter + "orderBy=" + args.orderBy;
            delimiter = "&";
        }
        if (args.orderDir) {
            searchURL += delimiter + "orderDir=" + args.orderDir;
            delimiter = "&";
        }
        if (args.page) {
            searchURL += delimiter + "page=" + args.page;
            delimiter = "&";
        }
        if (args.username) {
            searchURL += delimiter + "username=" + args.username;
        }
        // Send query and return response
        return this.get<Movie[]>(searchURL).then((response) => {
            return response as Movie[];
        });
    }

    /**
     * Likes a movie for a user.
     * Returns the updated status.
     *
     * @param {string} movieId - Id of the movie to like. e.g "tt9655334"
     * @param {string} username - The user to like the movie for
     * @return {Promise<Like>} Promise of a response with the updated like status
     */
    public likeMovie(movieId: string, username: string): Promise<Like> {
        return this.put<Like[]>(
            this.baseURL + "/user/" + username + "/LikedMovies/" + movieId
        ).then((response) => {
            return (response as Like[])[0];
        });
    }

    /**
     * Unlikes a movie for a user.
     * Returns the updated status.
     *
     * @param {string} movieId - Id of the movie to unlike. e.g "tt9655334"
     * @param {string} username - The user to unlike the movie for
     * @return {Promise<Like>} Promise of a response with the updated like status
     */
    public unlikeMovie(movieId: string, username: string): Promise<Like> {
        return this.delete<Like[]>(
            this.baseURL + "/user/" + username + "/LikedMovies/" + movieId
        ).then((response) => {
            return (response as Like[])[0];
        });
    }

    /**
     * Gets a list with all the liked movies for a user.
     *
     * @param {string} username - The user to get liked moves for
     * @return {Promise<Movie[]>} The promise of a list of the liked movies
     */
    public getLikedMovies(username: string): Promise<Movie[]> {
        return this.get<Movie[]>(
            this.baseURL + "/user/" + username + "/LikedMovies/"
        ).then((response) => {
            return response as Movie[];
        });
    }

    /**
     * Checks if the username already exists on the server.
     * If the username doesn't exist, request to create the user.
     * Return the user on success.
     *
     * @param {string} username - The username to check/create
     * @return {Promise<User>} The promise of a user
     */
    public loginOrCreateUser(username: string): Promise<User> {
        return new Promise<User>((resolve, reject) => {
            this.get<User[]>(this.baseURL + "/user/" + username)
                .then((users) => resolve(users[0] as User))
                .catch((error) => {
                    this.put<User[]>(this.baseURL + "/user/" + username)
                        .then((users) => {
                            resolve(users[0] as User);
                        })
                        .catch((error) => {
                            reject(error);
                        });
                });
        });
    }

    /**
     * Deletes a user.
     * Throws HttpRequestError if the user doesn't exist.
     * Returns the deleted user on success.
     *
     * @param {string} username - The username to delete
     * @return {Promise<User>} The promise of a deleted user
     */
    public deleteUser(username: string): Promise<User> {
        return this.delete<User[]>(this.baseURL + "/user/" + username).then(
            (response) => {
                return response[0] as User;
            }
        );
    }
}

// The url of the server
const baseURL = "http://it2810-22.idi.ntnu.no:3000";

// Construct and export the client
export default new HttpClient(baseURL);
