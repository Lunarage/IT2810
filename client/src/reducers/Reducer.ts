import {
    toggle_loggedIn,
    logout,
    set_username,
    setStoreSearchState,
} from "./Actions";
import { createStore, Store } from "redux";
import { Movie } from "../types/DatabaseTypes";

type Actions =
    | ReturnType<typeof toggle_loggedIn>
    | ReturnType<typeof logout>
    | ReturnType<typeof set_username>
    | ReturnType<typeof setStoreSearchState>;

// The type for holding search information and result
export type StoreSearchState = {
    inputSearch: string;
    inputOrderDir: "ASC" | "DESC";
    inputTitleType: string;
    pageActivePage: number;
    pageTotalPages: number;
    pageDisabled: boolean;
    searchStatus: "none" | "waiting" | "success" | "failure";
    searchErrorMessage: string | null;
    searchMovies: Movie[];
};

//Creates the types AppState
export type AppState = {
    loggedIn: boolean;
    userName?: string | null;
    searchState: StoreSearchState;
};

//Sets initial root state for the app
const initialAppState: AppState = {
    loggedIn: false,
    userName: null,
    searchState: {
        inputSearch: "",
        inputOrderDir: "ASC",
        inputTitleType: "",
        pageActivePage: 1,
        pageTotalPages: 2,
        pageDisabled: false,
        searchStatus: "none",
        searchErrorMessage: null,
        searchMovies: [],
    },
};
//Root reducer for the App, handels the states loggedIn/Loggedout and userName
export const rootReducer = (
    state: AppState = initialAppState,
    action: Actions
) => {
    switch (action.type) {
        case "TOGGLE_LOGGED_IN":
            return {
                ...state,
                loggedIn: action.payload,
            };

        case "SET_USERNAME":
            return {
                ...state,
                userName: action.payload,
            };

        case "LOGOUT":
            return {
                ...state,
                loggedIn: action.payload,
                userName: null,
            };
        case "SET_SEARCH_STATE":
            return {
                ...state,
                searchState: action.payload,
            };
        default:
            return state;
    }
};
//Configure store
function configureStore(): Store<AppState> {
    return createStore(
        rootReducer,
        (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
            (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    );
}

//Create and exports store
export const store = configureStore();
