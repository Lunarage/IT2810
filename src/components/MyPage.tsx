import React, {Component} from 'react';



class MyPage extends Component<{}, {}> {
    render() {
        return (
            <div className={"my-page"}>
                <h2 className={"my-page-title"}>{"Min side"}</h2>
                <h3>Lagrede søk</h3>
                <div className={"saved-searches"}>
                    <ul>Lagret søk nr 1</ul>
                    <ul>Lagret søk nr 2</ul>
                    <ul>Lagret søk nr 3</ul>
                    <ul>Lagret søk nr 4</ul>
                </div>
            </div>
        )
    }


}

export default MyPage;