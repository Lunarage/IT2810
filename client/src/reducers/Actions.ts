import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

export const toggle_loggedIn = (boolVal: boolean) => {
    return {
        type: "TOGGLE_LOGGED_IN",
        payload: boolVal
    } as const
};

export const set_username = ( text: string) => {
    return {
        type: "SET_USERNAME",
        payload: text
    } as const
};

export const logout = (boolVal: boolean) => {
    return {
        type: "LOGOUT",
        payload: boolVal
    } as const 
};
