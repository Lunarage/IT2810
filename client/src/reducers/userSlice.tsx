import { createSlice } from "@reduxjs/toolkit";
import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

// redux komponent som lager en "slice" av store hvor vi kan definere en reducer og actions som handler om en bruker logget inn eller ikke
const loginSlice = createSlice({
    name: "login/logout",
    initialState: { value: false, username: null },

    reducers: {
        logIn: (state, action) => {
            const baseURL = "http://it2810-22.idi.ntnu.no:3000";
            const client = new HttpClient(baseURL);
            const localStorage = new LocalStorageWrapper();
            localStorage.set("username", action.payload);
            client.createUser(action.payload);
            state.value = true;
            state.username = action.payload;
        },
        logOut: (state) => {
            const localStorage = new LocalStorageWrapper();
            localStorage.remove("username");
            state.value = false;
            state.username = null;
        },
    },
});

export const { logIn, logOut } = loginSlice.actions;

export default loginSlice.reducer;

/* These sites are used as src for this code: 
https://codesandbox.io/s/github/reduxjs/redux-essentials-counter-example/tree/master/?from-embed=&file=/src/features/counter/counterSlice.js:49-83
https://redux.js.org/tutorials/essentials/part-2-app-structure */
