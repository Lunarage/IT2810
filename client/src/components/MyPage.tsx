import React from "react";
import LikedMoviesTable from "./LikedMoviesTable";
import { useSelector, RootStateOrAny } from "react-redux";
import { AppState } from "../reducers/Reducer";

const Mypage = () => {
    const username = useSelector((state: AppState) => state.userName);

    const render = () => {
        if (username) {
            return (
                <div className={"my-page"}>
                    <h2 className={"my-page-title"}>{"Min side"}</h2>
                    <h3>Lagrede sÃ¸k</h3>
                    <LikedMoviesTable username={username} />
                </div>
            );
        } else {
            return <p>Hvordan kom du deg hit?</p>;
        }
    };
    return render();
};

export default Mypage;
