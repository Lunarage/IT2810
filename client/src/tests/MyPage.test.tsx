import { render, RenderResult } from "@testing-library/react";
import MyPage from "../components/MyPage";
import React from "react";
import { Provider } from "react-redux";
import { store } from "../reducers/Reducer";
import LikedMovieTable from "../components/LikedMoviesTable";
import HttpClient from "../modules/HttpClient";
import configureStore, {
    MockStoreCreator,
    MockStoreEnhanced,
} from "redux-mock-store";
import { act } from "react-dom/test-utils";

//Creates mock values when asking HTTP client for data
jest.mock("../modules/HttpClient");
const fakeHttpClient = HttpClient as jest.Mocked<typeof HttpClient>;
//Mocking request for getLikedMovies
fakeHttpClient.getLikedMovies.mockReturnValue(
    Promise.resolve([
        {
            tconst: "1",
            title_type: "Movie",
            primary_title: "Avengers",
            original_title: "Avengers",
            is_adult: true,
            end_year: null,
            start_year: 2012,
            runtime_minutes: 156,
            genres: "Action",
            liked: true,
        },
    ])
);
//Mocking request for getMovie
fakeHttpClient.getMovie.mockReturnValue(Promise.resolve({
  tconst: "1",
  title_type: "Movie",
  primary_title: "Avengers",
  original_title: "Avengers",
  is_adult: true,
  end_year: null,
  start_year: 2012,
  runtime_minutes: 156,
  genres: "Action",
  liked: true,
}))
//Checking if myPage renders
test("MyPage renders", () => {
    const { container } = render(
        <Provider store={store}>
            <MyPage />
        </Provider>
    );
    expect(container).toMatchSnapshot();
});
//Configure mockStore to be used later
const mockStore: MockStoreCreator = configureStore([]);

describe("MyPage", () => {
    let component: RenderResult;
    let store: MockStoreEnhanced;
    //run beforeEach test to insure correct state and component 
    beforeEach(() => {
        // Initialize mockStore with initial state
        store = mockStore({
            LoggedIn: true,
            userName: "testUser123",
        });

        component = render(
            <Provider store={store}>
                <MyPage />
            </Provider>
        );
    });

    it("should contain MyPage text", async () => {
        const MinSide = component.getByText("My page");
        expect(MinSide).toBeVisible()
    });
    
});
