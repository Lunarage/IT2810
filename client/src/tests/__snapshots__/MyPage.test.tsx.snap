// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders 1`] = `
<div>
  <p>
    Hvordan kom du deg hit?
  </p>
</div>
`;
